auth_pysaml/models/res_users.py

```diff
@@ -0,0 +1,209 @@
+                limit=1,
+            )
+            .user_id
+        )
+        if len(user) != 1:
+            raise AccessDenied()
+        # now find if a token for this user/provider already exists
+        token_ids = token_osv.search(
+            [("saml_provider_id", "=", provider), ("user_id", "=", user.id)]
+        )
+
+        if token_ids:
+            token_ids.write({"saml_access_token": saml_response})
+        else:
+            _logger.info("Creating auth_saml.token")
+            token_ids = token_osv.create(
+                {
+                    "saml_access_token": saml_response,
+                    "saml_provider_id": provider,
+                    "user_id": user.id,
+                }
+            )
+
+        if validation.get("mapped_attrs", {}):
+            user.write(validation.get("mapped_attrs", {}))
+
+        return user.login
+
+    @api.model
+    def auth_saml(self, provider, saml_response, base_url=None):
+        validation = self._auth_saml_validate(provider, saml_response, base_url)
+
+        # required check
+        if not validation.get("user_id"):
+            raise AccessDenied()
+
+        # retrieve and sign in user
+        login = self._auth_saml_signin(provider, validation, saml_response)
+
+        if not login:
+            raise AccessDenied()
+
+        # return user credentials
+        return self.env.cr.dbname, login, saml_response
+
+    def _check_credentials(self, token):
+        """Override to handle SAML auths.
+
+        The token can be a password if the user has used the normal form...
+        but we are more interested in the case when they are tokens
+        and the interesting code is inside the "except" clause.
+        """
+        try:
+            # Attempt a regular login (via other auth addons) first.
+            super(ResUser, self)._check_credentials(token)
+
+        except (AccessDenied, passlib.exc.PasswordSizeError):
+            # since normal auth did not succeed we now try to find if the user
+            # has an active token attached to his uid
+            res = (
+                self.env["auth_saml.token"]
+                .sudo()
+                .search(
+                    [
+                        ("user_id", "=", self.env.user.id),
+                        ("saml_access_token", "=", token),
+                    ]
+                )
+            )
+            if not res:
+                raise AccessDenied()
+
+    def _autoremove_password_if_saml(self):
+        """Helper to remove password if it is forbidden for SAML users."""
+        if self._allow_saml_and_password():
+            return
+        to_remove_password = self.filtered(
+            lambda rec: rec.id != SUPERUSER_ID
+            and rec.saml_ids
+            and not rec.password
+        )
+        to_remove_password.write(
+            {"password": pwd.genword(length=20, charset="ascii_72")}
```

Comment:
I think it should be cleared instead, otherwise the condition will be satified on every write

```suggestion
            {"password": False)}
```